### ISSUE TITLE
Inconsistent resampling behaviours across maskers

### ISSUE THREAD

**Issue by NicolasGensollen** [nilearn/nilearn#2720]
This issue came up when working on generalizing the reports for all maskers (see #2689). 
It seems like the different masker objects perform resampling of the mask and/or images at various moments of the fit-transform pipeline:

The `NiftiMasker` performs resampling explicitly in `fit`, meaning that the object itself is modified by the resampling. For example, here the `mask_img_` attribute is updated to match the target affine and shape:

https://github.com/nilearn/nilearn/blob/fa85bf0dd28687adc04743e95feecdc18c62bd17/nilearn/input_data/nifti_masker.py#L385-L389

However, after fitting, the object is not modified by calling `transform` although resampling will be done implicitly if needed in `filter_and_extract`.

The `NiftiLabelsMasker` performs resampling in `fit`, but also in `transform`, for example here the `_resampled_labels_img_` attribute is modified:

https://github.com/nilearn/nilearn/blob/fa85bf0dd28687adc04743e95feecdc18c62bd17/nilearn/input_data/nifti_labels_masker.py#L286-L295

This means that calling `transform` can potentially modify the object itself.

The `NiftiMapsMasker` also performs resampling in fit and `transform`:

https://github.com/nilearn/nilearn/blob/fa85bf0dd28687adc04743e95feecdc18c62bd17/nilearn/input_data/nifti_maps_masker.py#L302-L309

Finally, the `NiftiSpheresMasker` is never modified by resampling. It is only performed implicitly during the extraction:

https://github.com/nilearn/nilearn/blob/fa85bf0dd28687adc04743e95feecdc18c62bd17/nilearn/input_data/nifti_spheres_masker.py#L72-L77

I think we should aim for a consistent behaviour across maskers and avoid modifying maskers attributes in `transform`. WDYT?

**Comment by bthirion** [comment]
In general, I agree with this approach. 
I guess that there must be some reasons why mask resampling has been introduced in transform for `NiftiLabelsMasker`
@jeromedockes @GaelVaroquaux any hint ?

**Comment by Remi-Gau** [comment]
sklearn has a check that should enforce the masker object __dict__ to be unchanged by trasnform:

- see https://github.com/nilearn/nilearn/pull/5124

**Comment by Remi-Gau** [comment]
while working on #5124 I think we should try to follow what is enforced by sklearn estimator checks and not modify object when transform is called

if resampling needs to be done at transform time, it should be done without modifying the object

**Comment by bthirion** [comment]
Yes, it' would be good to be more consistent. It means that we may end up doing some mask and atlas resamplings several times (for each transform call) ----which could be very expensive in case of an atlas. Maybe this means that we should encourage providing samples at fit time so that the resampling gets done  once and for all there ?
